File: interop/shared/src/bootstrap.ts
/**
 * Walks a fresh rustak all the way to a credential a client could use.
 *
 * rustak has no local passwords and no "create an admin" flag: an installation
 * with nobody in it writes a one-time setup token to a file (0600) and the
 * first administrator is created with it, then immediately registers a passkey
 * — which is what actually hands back a bearer token
 * (`.claude/plan/status/M0-11-web-api-auth.md`). So this is the whole ceremony,
 * in the order an operator performs it:
 *
 *     GET  /api/v1/health                        wait for the listener
 *     GET  /api/v1/setup/status                  confirm nobody is here yet
 *     read <data_dir>/setup-token                the file the server wrote
 *     POST /api/v1/setup/admin                   → a registration token
 *     POST /api/v1/auth/passkey/register/start   → WebAuthn options
 *     POST /api/v1/auth/passkey/register/finish  → the administrator's session
 *     POST /api/v1/setup/server                  record the host name
 *     POST /api/v1/setup/complete                close the wizard for good
 *
 * From there a suite mints whatever it needs with [`createUser`],
 * [`setGroups`] and [`mintCredential`]: a client password for CloudTAK's
 * password grant, a one-time enrolment token for an EUD's `estream:`.
 *
 * The passkey step is why `webauthn.ts` exists. There is no non-browser path to
 * an administrator bearer token today, and no suite may add one — see
 * `interop/node-tak/README.md` → "The bootstrap gap".
 */

import fs from "node:fs";
import path from "node:path";

import { api, type HttpClient } from "./http.js";
import type { ServerInfo } from "./launch.js";
import { SoftAuthenticator, type CredentialOptions } from "./webauthn.js";

/** The administrator a bootstrap produced, and its bearer token. */
export interface Admin {
  readonly username: string;
  readonly token: string;
}

/** What a suite calls the installation it just created. */
export interface BootstrapOptions {
  /** The account the wizard creates. */
  readonly adminUsername: string;

  /** What to show beside it, and to label its passkey with. */
  readonly displayName: string;

  /** `[server] name`, recorded through the wizard. */
  readonly serverName: string;

  /** The domains the wizard records. The first is canonical. */
  readonly domains: readonly string[];

  /** The base URL the wizard records, which must be one of `domains`. */
  readonly baseUrl: string;
}

/** Waits until `/api/v1/health` answers, which is the first thing that can. */
export async function waitForHealth(client: HttpClient, timeoutMs = 60_000): Promise<void> {
  const deadline = Date.now() + timeoutMs;
  let last = "";

  while (Date.now() < deadline) {
    try {
      await api(client, "/health");
      return;
    } catch (error) {
      last = error instanceof Error ? error.message : String(error);
      await new Promise((resolve) => setTimeout(resolve, 200));
    }
  }

  throw new Error(`${client.base}/api/v1/health never answered. Last failure: ${last}`);
}

/**
 * Reads the one-time setup token the server wrote at first start.
 *
 * It is minted during start-up and the path is logged, so by the time the
 * health check answers it is there; the short retry is for the gap between the
 * listener binding and the file being flushed.
 */
async function readSetupToken(dataDir: string, timeoutMs = 10_000): Promise<string> {
  const file = path.join(dataDir, "setup-token");
  const deadline = Date.now() + timeoutMs;

  while (Date.now() < deadline) {
    try {
      const token = fs.readFileSync(file, "utf8").trim();

      if (token.length > 0) return token;
    } catch {
      // Not written yet.
    }

    await new Promise((resolve) => setTimeout(resolve, 100));
  }

  throw new Error(
    `${file} was never written, so there is no way to create the first administrator.`,
  );
}

/** Creates the administrator and signs it in with a freshly registered passkey. */
async function signIn(
  server: ServerInfo,
  client: HttpClient,
  options: BootstrapOptions,
): Promise<Admin> {
  const status = await api<{ needs_setup: boolean; has_admin: boolean }>(client, "/setup/status");

  if (!status.needs_setup || status.has_admin) {
    throw new Error("the scratch server already has an administrator, which cannot happen.");
  }

  const created = await api<{ username: string; registration_token: string }>(
    client,
    "/setup/admin",
    {
      body: {
        setup_token: await readSetupToken(server.dataDir),
        username: options.adminUsername,
        display_name: options.displayName,
      },
    },
  );

  const authenticator = new SoftAuthenticator(server.webtak);

  const challenge = await api<{ challenge_id: string; options: CredentialOptions }>(
    client,
    "/auth/passkey/register/start",
    { body: { label: options.displayName, registration_token: created.registration_token } },
  );

  const session = await api<{ token: string }>(client, "/auth/passkey/register/finish", {
    body: {
      challenge_id: challenge.challenge_id,
      credential: authenticator.create(challenge.options),
    },
  });

  return { username: created.username, token: session.token };
}

/** Records the host name and closes the wizard, as an operator would. */
async function finishWizard(
  client: HttpClient,
  token: string,
  options: BootstrapOptions,
): Promise<void> {
  await api(client, "/setup/server", {
    body: { name: options.serverName, domains: options.domains, base_url: options.baseUrl },
    token,
  });

  await api(client, "/setup/complete", { method: "POST", token });
}

/** Performs the whole walk against a server that has just started. */
export async function bootstrap(
  server: ServerInfo,
  client: HttpClient,
  options: BootstrapOptions,
): Promise<Admin> {
  await waitForHealth(client);

  const admin = await signIn(server, client, options);

  await finishWizard(client, admin.token, options);

  const me = await api<{ username: string; is_admin: boolean }>(client, "/me", {
    token: admin.token,
  });

  if (me.username !== admin.username || !me.is_admin) {
    throw new Error(`the bootstrap signed in as ${me.username}, which is not an administrator.`);
  }

  return admin;
}

/** Creates an ordinary account, and returns the username the server settled on. */
export async function createUser(
  client: HttpClient,
  token: string,
  username: string,
  displayName: string,
): Promise<string> {
  const created = await api<{ username: string }>(client, "/users", {
    body: { username, display_name: displayName },
    token,
  });

  return created.username;
}

/** What minting handed back: the only moment the secret exists outside the server. */
export interface MintedCredential {
  readonly id: number;
  readonly secret: string;

  /** The `tak://` URL an ATAK user scans, for an enrolment token. */
  readonly enrollUrl?: string;
}

/** Mints a credential for `username`, as an administrator may on anyone's behalf. */
export async function mintCredential(
  client: HttpClient,
  token: string,
  request: {
    kind: "enrollment_token" | "client_password";
    label: string;
    username: string;
    expiresInDays?: number;
    maxUses?: number;
  },
): Promise<MintedCredential> {
  const minted = await api<{
    credential: { id: number };
    secret: string;
    enroll_url?: string;
  }>(client, "/credentials", {
    body: {
      kind: request.kind,
      label: request.label,
      username: request.username,
      ...(request.expiresInDays === undefined ? {} : { expires_in_days: request.expiresInDays }),
      ...(request.maxUses === undefined ? {} : { max_uses: request.maxUses }),
    },
    token,
  });

  return { id: minted.credential.id, secret: minted.secret, enrollUrl: minted.enroll_url };
}

/** One channel grant, in the shape `PUT /api/v1/users/{username}/groups` takes. */
export interface Grant {
  readonly group: string;

  /** TAK's own vocabulary: `IN` is permission to write, `OUT` to read. */
  readonly direction: "IN" | "OUT" | "BOTH";
}

/** Creates a channel, tolerating one that is already there. */
export async function createGroup(
  client: HttpClient,
  token: string,
  name: string,
): Promise<void> {
  const response = await client.request("/api/v1/groups", {
    body: { name },
    token,
  });

  // 409 is "somebody already made it", which is the state this asked for.
  if (response.status === 409 || (response.status >= 200 && response.status < 300)) return;

  throw new Error(`POST /api/v1/groups answered ${response.status}: ${response.body}`);
}

/** Replaces the channels an account is granted by hand. */
export async function setGroups(
  client: HttpClient,
  token: string,
  username: string,
  grants: readonly Grant[],
): Promise<void> {
  await api(client, `/users/${encodeURIComponent(username)}/groups`, {
    method: "PUT",
    body: grants.map((grant) => ({ group: grant.group, direction: grant.direction })),
    token,
  });
}

File: rustak-server/tests/bootstrap.rs
//! The server, started the way `main` starts it.
//!
//! Everything else in this crate tests a piece: a route through
//! `actix_web::test`, a repository against an in-memory database, a ceremony
//! against the software authenticator. None of that proves the pieces are
//! wired together, and the failures that matter most at start-up are exactly
//! the wiring ones — a certificate that is never issued, a handle installed too
//! late, a shutdown that hangs. So this suite drives [`rustak_server::run`] over
//! real sockets: a temporary data directory, a listener presenting a
//! certificate from an authority that did not exist a second ago, an HTTP
//! client that trusts it, and the whole first-run wizard through to a session
//! that `/api/v1/me` accepts.
//!
//! # Why the client is told how to resolve the host
//!
//! WebAuthn identifies a relying party by domain, so `https://127.0.0.1` cannot
//! register a passkey at all. The server is therefore reached at `localhost`,
//! and `reqwest`'s own resolver override points that at the loopback address
//! the listener bound — which also means these tests never depend on what the
//! machine's `/etc/hosts` says, or on whether `localhost` resolves to IPv6
//! first.
//!
//! Run with `cargo test -p rustak-server --features testing`.

#![cfg(feature = "testing")]

use std::net::SocketAddr;
use std::path::Path;
use std::sync::Arc;
use std::time::Duration;

use rustak_api::{AdminCreated, Me, PasskeyChallenge, SetupStatus, TokenResponse};
use rustak_core::config::ListenAddr;
use rustak_core::runtime::Shutdown;
use rustak_core::telemetry::Session;
use rustak_server::config::{Config, TlsMode};
use rustak_server::testing::SoftAuthenticator;
use tokio::task::JoinHandle;

/// How long a test waits for the listener to come up.
///
/// Generous, because a first start generates an RSA certificate authority and
/// a signing key, which on a loaded CI runner is seconds rather than
/// milliseconds.
const STARTUP_TIMEOUT: Duration = Duration::from_secs(30);

/// How long a cancelled server is given to return, per design 01 §8 step 12.
const SHUTDOWN_TIMEOUT: Duration = Duration::from_secs(10);

/// A server that has stopped, and the data directory it left behind.
struct Stopped {
    outcome: Result<(), human_errors::Error>,
    wal: std::path::PathBuf,
    data_dir: tempfile::TempDir,
}

impl Stopped {
    /// Asserts that the server returned `Ok(())` and folded its log back in.
    fn assert_clean(self) {
        self.outcome.expect("a cancelled server returns Ok");

        let wal = self
            .wal
            .metadata()
            .map(|meta| meta.len())
            .unwrap_or_else(|err| panic!("the -wal file should still be there: {err}"));

        assert_eq!(
            wal, 0,
            "the write-ahead log should have been truncated on the way out",
        );

        drop(self.data_dir);
    }
}

/// A server running in this process, and what is needed to talk to it.
struct Running {
    shutdown: Shutdown,
    handle: JoinHandle<Result<(), human_errors::Error>>,
    client: reqwest::Client,
    base: String,
    data_dir: tempfile::TempDir,
    config: Config,
}

impl Running {
    /// The full URL of a path on the server under test.
    fn url(&self, path: &str) -> String {
        format!("{}{path}", self.base)
    }

    /// Cancels the server and waits for `run` to return.
    ///
    /// The client goes first, because actix drains rather than cuts: an idle
    /// keep-alive connection — and over TLS that is an HTTP/2 one, because the
    /// listener advertises `h2` — holds the drain open for the full
    /// `shutdown_timeout`. A browser that is still connected really would cost
    /// those ten seconds, so this closes the connection the way a client
    /// leaving does, and what is measured afterwards is the shutdown rather
    /// than that timeout.
    async fn stop(self) -> Stopped {
        drop(self.client);
        self.shutdown.cancel();

        let outcome = tokio::time::timeout(SHUTDOWN_TIMEOUT, self.handle)
            .await
            .expect("a cancelled server returns within ten seconds")
            .expect("the server task should not panic");

        Stopped {
            outcome,
            wal: wal_path(&self.config),
            // Returned rather than dropped, because dropping it deletes the
            // data directory — and a test that asserted on a file inside one
            // that had already gone would pass by reading nothing.
            data_dir: self.data_dir,
        }
    }
}

/// A port nothing is listening on.
///
/// The socket is closed before the server binds it, which is a race in
/// principle; in practice the operating system does not hand the same ephemeral
/// port out twice in the moment between, and the alternative — asking the
/// server which port it ended up on — would be API surface that exists only for
/// this file.
fn free_port() -> u16 {
    std::net::TcpListener::bind("127.0.0.1:0")
        .expect("a free loopback port")
        .local_addr()
        .expect("the bound address")
        .port()
}

/// A telemetry session that records into memory.
fn session() -> Arc<Session> {
    Arc::new(Session::new("rustak", "0.0.0-test").with_battery(tracing_batteries::Testing))
}

/// Starts a server on a port of its own and waits until it answers.
///
/// `tls` chooses between the internal authority and the plaintext listener the
/// insecure development mode allows, which are the two shapes `[web.public]`
/// can take in M0.
async fn start(tls: bool) -> Running {
    let data_dir = tempfile::tempdir().expect("a temporary data directory");
    let port = free_port();
    let address: SocketAddr = format!("127.0.0.1:{port}")
        .parse()
        .expect("a loopback address");

    let mut config = Config::testing(data_dir.path());
    config.web.public.listen = vec![ListenAddr::new("127.0.0.1", port)];
    config.server.base_url = Some(format!(
        "{}://localhost:{port}",
        if tls { "https" } else { "http" }
    ));

    if tls {
        config.web.public.tls.mode = TlsMode::Internal;
        config.web.public.allow_insecure_http = false;
    }

    let shutdown = Shutdown::new();
    let handle = tokio::spawn(rustak_server::run(
        config.clone(),
        session(),
        shutdown.clone(),
    ));

    let base = config
        .server
        .base_url
        .clone()
        .expect("the base URL just configured");

    let client = client(tls, config.server.data_dir.as_path(), address).await;
    await_listener(&client, &base).await;

    Running {
        shutdown,
        handle,
        client,
        base,
        data_dir,
        config,
    }
}

/// An HTTP client that trusts this installation's own authority and resolves
/// `localhost` to the socket the listener bound.
async fn client(tls: bool, data_dir: &Path, address: SocketAddr) -> reqwest::Client {
    let mut builder = reqwest::Client::builder()
        .resolve("localhost", address)
        .timeout(Duration::from_secs(20));

    if tls {
        // The authority is created during start-up, so it is not on disk yet
        // when this is called — which is also the first thing this suite
        // proves, because a client that cannot find it cannot trust it.
        let ca = await_file(&rustak_server::pki::ca_certificate_path(data_dir)).await;

        builder = builder.add_root_certificate(
            reqwest::Certificate::from_pem(&ca).expect("the exported authority is valid PEM"),
        );
    }

    builder.build().expect("an HTTP client")
}

/// Waits for a file start-up writes, and reads it.
async fn await_file(path: &Path) -> Vec<u8> {
    let deadline = std::time::Instant::now() + STARTUP_TIMEOUT;

    loop {
        if let Ok(contents) = tokio::fs::read(path).await {
            return contents;
        }

        assert!(
            std::time::Instant::now() < deadline,
            "{} was never written",
            path.display(),
        );

        tokio::time::sleep(Duration::from_millis(25)).await;
    }
}

/// Waits until the listener answers, rather than guessing how long it takes.
async fn await_listener(client: &reqwest::Client, base: &str) {
    let deadline = std::time::Instant::now() + STARTUP_TIMEOUT;

    loop {
        let last = match client.get(format!("{base}/api/v1/health")).send().await {
            Ok(response) if response.status().is_success() => return,
            Ok(response) => response.status().to_string(),
            Err(err) => err.to_string(),
        };

        assert!(
            std::time::Instant::now() < deadline,
            "the listener never came up: {last}",
        );

        tokio::time::sleep(Duration::from_millis(25)).await;
    }
}

/// The body of a successful request, as JSON.
async fn get_json<T: serde::de::DeserializeOwned>(server: &Running, path: &str) -> T {
    let response = server
        .client
        .get(server.url(path))
        .send()
        .await
        .unwrap_or_else(|err| panic!("GET {path}: {err}"));

    assert!(response.status().is_success(), "GET {path}: {:?}", response);

    response.json().await.expect("a JSON body")
}

/// Posts JSON and asserts the request succeeded.
async fn post_json<T: serde::de::DeserializeOwned>(
    server: &Running,
    path: &str,
    body: serde_json::Value,
    bearer: Option<&str>,
) -> T {
    let mut request = server.client.post(server.url(path)).json(&body);

    if let Some(token) = bearer {
        request = request.bearer_auth(token);
    }

    let response = request
        .send()
        .await
        .unwrap_or_else(|err| panic!("POST {path}: {err}"));
    let status = response.status();
    let text = response.text().await.expect("a response body");

    assert!(status.is_success(), "POST {path} answered {status}: {text}");

    serde_json::from_str(&text).unwrap_or_else(|err| panic!("POST {path} returned {text}: {err}"))
}

#[tokio::test(flavor = "multi_thread")]
async fn a_first_start_serves_its_own_tls_and_walks_an_operator_all_the_way_in() {
    let server = start(true).await;

    // The embedded UI, over a certificate issued minutes ago by an authority
    // that did not exist before this test ran.
    let robots = server
        .client
        .get(server.url("/robots.txt"))
        .send()
        .await
        .expect("GET /robots.txt");
    assert!(robots.status().is_success());
    assert!(robots.text().await.unwrap().contains("Disallow"));

    let health: serde_json::Value = get_json(&server, "/api/v1/health").await;
    assert_eq!(health["status"], "ok", "{health}");

    // The wizard, from the token on disk to a session.
    let status: SetupStatus = get_json(&server, "/api/v1/setup/status").await;
    assert!(status.needs_setup && !status.has_admin);

    let token = String::from_utf8(await_file(&server.config.setup_token_file()).await)
        .expect("the setup token is text");

    let created: AdminCreated = post_json(
        &server,
        "/api/v1/setup/admin",
        serde_json::json!({
            "setup_token": token.trim(),
            "username": "ada",
            "display_name": "Ada Lovelace",
        }),
        None,
    )
    .await;
    assert_eq!(created.username.as_str(), "ada");

    let challenge: PasskeyChallenge = post_json(
        &server,
        "/api/v1/auth/passkey/register/start",
        serde_json::json!({
            "label": "Laptop",
            "registration_token": created.registration_token,
        }),
        None,
    )
    .await;

    let authenticator = SoftAuthenticator::new(&server.base);
    let credential = authenticator.create(&challenge.options);

    // The wizard's registration answers with a session, because the account it
    // just created has no other way to authenticate.
    let signed_in: TokenResponse = post_json(
        &server,
        "/api/v1/auth/passkey/register/finish",
        serde_json::json!({
            "challenge_id": challenge.challenge_id,
            "credential": credential,
        }),
        None,
    )
    .await;

    let me: Me = server
        .client
        .get(server.url("/api/v1/me"))
        .bearer_auth(&signed_in.token)
        .send()
        .await
        .expect("GET /api/v1/me")
        .json()
        .await
        .expect("a JSON identity");

    assert_eq!(me.username.as_str(), "ada");
    assert!(me.is_admin, "the wizard creates an administrator");

    // The whole point of closing the database rather than dropping it: what is
    // left behind is a database and an empty log, not megabytes of pending
    // writes somebody's backup would miss.
    server.stop().await.assert_clean();
}

#[tokio::test(flavor = "multi_thread")]
async fn the_insecure_development_listener_serves_plaintext_and_still_stops_cleanly() {
    // The other shape `[web.public]` can take, and the one every other test in
    // this crate runs under — worth proving it binds a real socket too.
    let server = start(false).await;

    let robots = server
        .client
        .get(server.url("/robots.txt"))
        .send()
        .await
        .expect("GET /robots.txt");
    assert!(robots.status().is_success());

    let started = std::time::Instant::now();
    server.stop().await.assert_clean();

    assert!(
        started.elapsed() < SHUTDOWN_TIMEOUT,
        "a server with nothing in flight should stop promptly",
    );
}

#[tokio::test(flavor = "multi_thread")]
async fn a_listener_that_cannot_bind_reports_it_rather_than_running_without_one() {
    // A server that came up with no socket would look healthy in an
    // orchestrator and answer nothing, so the failure has to reach the caller.
    let data_dir = tempfile::tempdir().expect("a temporary data directory");
    let occupied = std::net::TcpListener::bind("127.0.0.1:0").expect("a port to occupy");
    let port = occupied.local_addr().unwrap().port();

    let mut config = Config::testing(data_dir.path());
    config.web.public.listen = vec![ListenAddr::new("127.0.0.1", port)];

    let Err(err) = rustak_server::run(config, session(), Shutdown::new()).await else {
        panic!("binding a port somebody else holds should not start a server");
    };

    assert!(err.is(human_errors::Kind::User), "{err}");
    assert!(err.to_string().contains(&port.to_string()), "{err}");
}

/// Where the write-ahead log lives for a configured server.
fn wal_path(config: &Config) -> std::path::PathBuf {
    let database = config.database_path();
    let mut name = database.file_name().unwrap_or_default().to_os_string();
    name.push("-wal");

    database.with_file_name(name)
}
